rules/chaos.go

```diff
@@ -23,18 +28,34 @@ func init() {
 }
 
 func (rule *chaos) load() (bool, string, error) {
-	value, active := os.LookupEnv(envChaosChance)
-	if !active {
+	explicit := explicitLoad(ruleChaos)
+	value, hasDefault := os.LookupEnv(envChaosChance)
+	if !explicit && !hasDefault {
 		return false, "", nil
 	}
 	chance, err := strconv.ParseFloat(value, 64)
-	if err != nil {
+	if !explicit && err != nil {
 		return false, "", fmt.Errorf("invalid chaos chance %s", err)
 	}
 	rule.chance = chance
-	return true, fmt.Sprintf("chaos chance %s", value), nil
+
+	if rule.chance != 0 {
+		return true, fmt.Sprintf("chaos chance %s", value), nil
+	}
+	return true, fmt.Sprint("chaos (no default)"), nil
```

Comment:
I want to make sure I understand the desired behavior here, as there are 4 possible cases based on explicit load and default value.

1. NO explicit load and NO default value -- don't load the rule
2. NO explicit load but default value -- load the rule (previous behavior) -- this case honors the annotation overrides?
3. explicit load but NO default value -- load the rule, but the rule would also say "don't reap" unless the pod had an annotation override and the condition for that override was met?
4. explicit load but default value -- loads the rule and and default value, but honors the annotation overrides.

Wondering if we need a way to determine whether or not the reaper should honor the annotations? I guess the concern is making sure that if the reaper isn't intended to be overwritten that it can't be. Maybe I'm missing that piece here (I'm still looking!)

There might also be a case where explicit loading of the rule is set but there's also a value that isn't parsible as a float (i.e. `lizard`) here?

Reply:
You are correct for all 4 cases. If a rule is loaded (explicitly or with a default value), it will always check for an annotation and use the annotation's value if found. I don't think we need an option to explicitly enable annotations. It should be safe to assume that if a user added a `pod-reaper/*` annotation, they intend to opt-in to annotation override. If a pod is not annotated, then pod-reaper will behave the same as it does now.

The annotation is parsed in `ShouldReap()`. If a pod is annotated with `pod-reaper/chaos-chance: lizard`, a warning will be logged, and the behavior will fall back to default value.

Reply:
The chaos rule might be a weird edge case since all enabled rules must return true. If you were to enable chaos without a default, it would prevent any pod which was not annotated from being reaped. I would prefer that a pod is reaped if _any_ rule returns true, but that would be a potentially breaking change. We could also implement reap/spare logic like you mentioned in https://github.com/target/pod-reaper/issues/44#issuecomment-586729946, but I wanted to keep things simple if possible.

Reply:
yeah, reaping if *any* rule returns true would be breaking. If it operator by reaping when any rule returned true there wouldn't be a way to get an "or" function between multiple rules. How I've done the "reaping on any rule = true" situation was to make multiple reapers, each with only one rule.